app/code/core/Mage/Core/Model/Config.php

```diff
@@ -501,13 +491,44 @@ public function getCache()
     }
 
     /**
-     * Get lock flag cache identifier
+     * Call before building and saving cache to ensure only one process can save the cache
      *
-     * @return string
+     * If failed to get cache lock:
+     *   - CLI: throws exception
+     *   - Other: 503 error
+     *
+     * @return void
+     * @throws Exception
      */
-    protected function _getCacheLockId()
+    public function getCacheSaveLock()
     {
-        return $this->getCacheId() . '.lock';
+        if (!Mage::app()->useCache('config')) {
+            return;
+        }
+        $connection = Mage::getSingleton('core/resource')->getConnection('core_write');
+        $lockWaitTime = getenv('MAGE_CONFIG_CACHE_LOCK_WAIT') ?: (PHP_SAPI === 'cli' ? 60 : 3);
+        if (!$connection->fetchOne("SELECT GET_LOCK('core_config_cache_save_lock', ?)", [$lockWaitTime])) {
+            if (PHP_SAPI === 'cli') {
+                throw new Exception('Could not get lock on cache save operation.');
+            } else {
+                require Mage::getBaseDir() . DS . 'errors' . DS . '503.php';
+                die();
+            }
+        }
+    }
+
+    /**
+     * Release the cache saving lock after it is saved or no longer needed
+     *
+     * @return void
+     */
+    public function releaseCacheSaveLock()
+    {
+        if (!Mage::app()->useCache('config')) {
+            return;
+        }
+        $connection = Mage::getSingleton('core/resource')->getConnection('core_write');
+        $connection->fetchOne("SELECT RELEASE_LOCK('core_config_cache_save_lock')");
```

Comment:
I would put the name of lock `core_config_cache_save_lock` to constant, to avoid magic-string in unit tests.

Reply:
What unit tests? 😉

Reply:
Future 🙂